[inductor][user triton] Check size hints to determine indexing dtype

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137234

Previously, all integer inputs to user-defined triton kernels were assumed to be int32. This would result in errors if your input was actually an int64.

This PR checks the value to determine which dtype to use for indexing: if it is known to be < int_max, then use int32 (and add guards if relevant); if we can't check (e.g. unbacked symint), then use int64.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @oulgen @aakhundov

Differential Revision: [D63797975](https://our.internmc.facebook.com/intern/diff/D63797975)

davidberard98 committed:
[inductor][user triton] Check size hints to determine tl.int32 vs tl.int64 indexing

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63797975).

davidberard98 committed:
Update on "[inductor][user triton] Check size hints to determine tl.int32 vs tl.int64 indexing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D63797975](https://our.internmc.facebook.com/intern/diff/D63797975)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63797975).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63797975).

aakhundov commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -53,10 +53,20 @@ def signature_of(arg: KernelArgType, *, size_dtype: str) -> str:
         elif isinstance(arg.expr, (float, sympy.Float)):
             return "fp32"
+
+        # if this is a integer
```
Super-nit: `an integer` :)

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1677,6 +1677,47 @@ def fn(x):
+
+        def fn(x):
+            x = x * x
+            seed = torch.randint(
+                low=0, high=2**62, size=(1,), dtype=torch.int64
+            ).item()
```
Do we want to explicitly put this in the above int32 range, so that we don't get errors passing silently? Or maybe check both within and outside int32?

davidberard98 replied:
Good point, updated it to `[2**32, 2**62)`.

aakhundov replied:
Thanks for doing this, although I realized how insanely unlikely it is sample int32 from [2, 2**62] :)

davidberard98 committed:
Update on "[inductor][user triton] Check size hints to determine indexing dtype"


Previously, all integer inputs to user-defined triton kernels were assumed to be int32. This would result in errors if your input was actually an int64.

This PR checks the value to determine which dtype to use for indexing: if it is known to be < int_max, then use int32 (and add guards if relevant); if we can't check (e.g. unbacked symint), then use int64.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang oulgen aakhundov

Differential Revision: [D63797975](https://our.internmc.facebook.com/intern/diff/D63797975)

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][user triton] Check size hints to determine indexing dtype"


Previously, all integer inputs to user-defined triton kernels were assumed to be int32. This would result in errors if your input was actually an int64.

This PR checks the value to determine which dtype to use for indexing: if it is known to be < int_max, then use int32 (and add guards if relevant); if we can't check (e.g. unbacked symint), then use int64.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang oulgen aakhundov

Differential Revision: [D63797975](https://our.internmc.facebook.com/intern/diff/D63797975)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63797975).

davidberard98 commented:
@pytorchbot merge
